src/status.ts

```diff
@@ -2,258 +2,396 @@ import * as vscode from 'vscode';
-    this._reloadActiveFolderButton();
+  protected getTextNormal():string {
+    return this._text;
+  }
+  protected getTextShort():string {
+    return this.getTextNormal();
+  }
+  protected getTextIcon():string {
+    return this.getTextShort();
   }
+}
+
+class ActiveFolderButton extends Button {
+  settingsName = 'workspace';
+  command = "cmake.selectActiveFolder";
 
-  private _autoSelectActiveFolder: boolean = false;
-  setAutoSelectActiveFolder(autoSelectActiveFolder: boolean) {
-    this._autoSelectActiveFolder = autoSelectActiveFolder;
-    this._reloadActiveFolderButton();
+  private static readonly _autoSelectToolTip = localize('active.folder.auto.select.tooltip', 'Active folder');
+  private static readonly _toolTip = localize('active.folder.tooltip', 'Select Active folder');
+  private _autoSelect: boolean = false;
+
+  set autoSelect(v:boolean) {
+    this._autoSelect = v;
+    this.update();
   }
 
-  /**
-   * The build type label. Determined by the active build variant
-   */
-  private _buildTypeLabel: string = localize('unconfigured', 'Unconfigured');
-  setBuildTypeLabel(v: string) {
-    this._buildTypeLabel = v;
-    this._reloadStatusButton();
+  protected getTooltipNormal():string|null {
+    if (this._autoSelect) {
+      return ActiveFolderButton._autoSelectToolTip;
+    }
+    return ActiveFolderButton._toolTip;
+  }
+  protected isVisible():boolean {
+    return Boolean(vscode.workspace.workspaceFolders && vscode.workspace.workspaceFolders.length > 1 && !!this.getText());
+  }
+  protected getTextIcon():string {
+    return '$(folder-active)';
   }
+  protected getTooltipIcon():string {
+    return `CMake: ${this.getTooltipNormal()}\n${this.getTextNormal()}`;
+  }
+}
+
+class CMakeStatus extends Button {
+  settingsName = 'status';
+  command = "cmake.setVariant";
+  tooltip = localize('click.to.select.variant.tooltip', 'Click to select the current build variant');
 
-  /**
-   * The message shown in the primary status button. Tells the user what the
-   * extension is currently up to.
-   */
+  private _buildTypeLabel: string = localize('unconfigured', 'Unconfigured');
   private _statusMessage: string = localize('loading.status', 'Loading...');
-  setStatusMessage(v: string) {
+
+  set buildTypeLabel(v: string) {
+    this._buildTypeLabel = v;
+    this.update();
+  }
+  set statusMessage(v: string) {
     this._statusMessage = v;
-    this._reloadStatusButton();
+    this.update();
   }
-
-  /**
-   * The name of the currently active target to build
-   */
-  private _targetName: string = '';
-  public get targetName(): string { return this._targetName; }
-  public set targetName(v: string) {
-    this._targetName = v;
-    this._targetButton.text = `[${v}]`;
-    this.reloadVisibility();
+  protected getTextNormal() {
+    return `CMake: ${this._buildTypeLabel}: ${this._statusMessage}`;
+  }
+  protected getTextShort() {
+    return `${this._buildTypeLabel}: ${this._statusMessage}`;
   }
+}
```

Comment:
I think the text for the cmake status label should disappear in "icon" mode

Reply:
now it is:
text -> CMake: type: status
short -> type: status
icon -> hidden

i'm thinking about adding an icon, but i'm not sure which one to use.

I also noticed, that the status message can be quite big. maybe it would be a good idea to move the status message into the tooltip.

Reply:
`$(info)` might be an appropriate icon maybe?  I'm not sold on keeping the status the way it is right now, but that's not something you need to solve in this PR.  In my development, I only use it to select variants, not to look at status.

Reply:
i added the icon and removed the status info from the label in compact mode.